incubator/orm/indexer.go

```diff
@@ -0,0 +1,119 @@
+package orm
+
+import (
+	"encoding/binary"
+	"math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// IndexerFunc creates one or multiple index keys for the source object.
+type IndexerFunc func(value interface{}) ([][]byte, error)
+
+// Indexer manages the persistence for an index based on searchable keys and operations.
+type Indexer struct {
+	indexerFunc IndexerFunc
+	addPolicy   func(store sdk.KVStore, secondaryIndexKey []byte, rowId uint64) error
+}
+
+func NewIndexer(indexerFunc IndexerFunc, unique bool) *Indexer {
+	i := &Indexer{indexerFunc: indexerFunc}
+	if unique {
+		i.addPolicy = uniqueKeysAddPolicy
+	} else {
+		i.addPolicy = multiKeyAddPolicy
+	}
+	return i
+}
+
+func (u Indexer) OnCreate(store sdk.KVStore, rowId uint64, value interface{}) error {
+	secondaryIndexKeys, err := u.indexerFunc(value)
+	if err != nil {
+		return err
+	}
+
+	for _, secondaryIndexKey := range secondaryIndexKeys {
+		if err := u.addPolicy(store, secondaryIndexKey, rowId); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (u Indexer) OnDelete(store sdk.KVStore, rowId uint64, value interface{}) error {
+	secondaryIndexKeys, err := u.indexerFunc(value)
+	if err != nil {
+		return err
+	}
+	for _, secondaryIndexKey := range secondaryIndexKeys {
+		indexKey := makeIndexPrefixScanKey(secondaryIndexKey, rowId)
+		store.Delete(indexKey)
+	}
+	return nil
+}
+
+func (u Indexer) OnUpdate(store sdk.KVStore, rowId uint64, newValue, oldValue interface{}) error {
+	oldSecIdxKeys, err := u.indexerFunc(oldValue)
```

Comment:
Cool. I was thinking this whole difference is overkill, but then I realized that indexerFunc() returns `[][]byte`. So it could be 10 users, updated to 5 and 3 overlap.